<a href="https://colab.research.google.com/github/elerufe/DiploDatosFaMAF/blob/main/EyCD/G25_Entregable_EyCD_parte_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Grupo 25 - Entregable EyCD

#Parte 1 de 2

Integrantes:

Rufeil Fiori, Elena

Sosa, Manuel

Taié Mirón, Macarena

Yurovich, Luis Federico

**Diplomatura en Ciencia de Datos, Aprendizaje Automático y sus Aplicaciones**

**Exploración y Curación de Datos**

*Edición 2022*

----

# Trabajo práctico entregable - parte 1

En esta notebook, vamos a cargar el conjunto de datos de [la compentencia Kaggle](https://www.kaggle.com/dansbecker/melbourne-housing-snapshot) sobre estimación de precios de ventas de propiedades en Melbourne, Australia.

Utilizaremos el conjunto de datos reducido producido por [DanB](https://www.kaggle.com/dansbecker). Hemos subido una copia a un servidor de la Universidad Nacional de Córdoba para facilitar su acceso remoto.

In [ ]:
import matplotlib.pyplot as plt
import numpy
import pandas

import seaborn
seaborn.set_context('talk')

!pip install python-decouple
#from decouple import config
from sqlalchemy import create_engine, text

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import plotly
plotly.__version__
# Make sure it's 4.14.3

'5.5.0'

In [ ]:
melbourne_df = pandas.read_csv(
    'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/melb_data.csv')
melb_df = melbourne_df.copy()
melb_df[:3]

Suburb          Address  Rooms Type      Price Method SellerG  \
0  Abbotsford     85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford     5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  

[3 rows x 21 columns]

Primero inspeccionamos las columnas del dataframe

In [ ]:
melb_df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

Aqui observamos la cantidad de registros.

In [ ]:
len(melb_df)

13580

Cargamos la base de datos de airbnb.

In [ ]:
airbnb_df = pandas.read_csv('https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/cleansed_listings_dec18.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (35,77) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [ ]:
airbnb_df[:3]

id                         listing_url     scrape_id last_scraped  \
0   9835   https://www.airbnb.com/rooms/9835  2.018120e+13    12/7/2018   
1  10803  https://www.airbnb.com/rooms/10803  2.018120e+13    12/7/2018   
2  12936  https://www.airbnb.com/rooms/12936  2.018120e+13    12/7/2018   

                                               name  \
0                            Beautiful Room & House   
1          Room in Cool Deco Apartment in Brunswick   
2  St Kilda 1BR APT+BEACHSIDE+VIEWS+PARKING+WIFI+AC   

                                             summary  \
0                                                NaN   
1  A large air conditioned room with queen spring...   
2  RIGHT IN THE HEART OF ST KILDA! It doesn't get...   

                                               space  \
0  House: Clean, New, Modern, Quite, Safe. 10Km f...   
1  The apartment is Deco/Edwardian in style and h...   
2  FREE WiFi FREE in-building remote controlled g...   

                                         description  \
0  House: Clean, New, Modern, Quite, Safe. 10Km f...   
1  A large air conditioned room with queen spring...   
2  RIGHT IN THE HEART OF ST KILDA! It doesn't get...   

                               neighborhood_overview  \
0       Very safe! Family oriented. Older age group.   
1  This hip area is a crossroads between two grea...   
2  A stay at our apartment means you can enjoy so...   

                                               notes  ...  \
0                                                NaN  ...   
1                                                NaN  ...   
2  First floor apartment with both lift and stair...  ...   

  review_scores_location review_scores_value requires_license license  \
0                    9.0                 9.0                f     NaN   
1                    9.0                 9.0                f     NaN   
2                    9.0                 9.0                f     NaN   

  instant_bookable          cancellation_policy require_guest_profile_picture  \
0                f  strict_14_with_grace_period                             f   
1                t                     moderate                             t   
2                f  strict_14_with_grace_period                             f   

  require_guest_phone_verification calculated_host_listings_count  \
0                                f                              1   
1                                t                              1   
2                                f                             17   

  reviews_per_month  
0              0.04  
1              1.50  
2              0.15  

[3 rows x 84 columns]

In [ ]:
print(airbnb_df.isnull().sum()[:50])
print(airbnb_df.isnull().sum()[50:])

id                            0
listing_url                   0
scrape_id                     0
last_scraped                  0
name                          3
summary                     696
space                      6051
description                 332
neighborhood_overview      8471
notes                     11349
transit                    7952
access                     7727
interaction                8358
house_rules                7863
picture_url                   0
host_id                       0
host_url                      0
host_name                     3
host_since                    3
host_location                26
host_about                 9165
host_response_time         7306
host_response_rate         7306
host_is_superhost             3
host_thumbnail_url            3
host_picture_url              3
host_neighborhood          7968
host_verifications            0
host_has_profile_pic          3
host_identity_verified        3
street                        0
neighbor

In [ ]:
airbnb_df.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'summary',
       'space', 'description', 'neighborhood_overview', 'notes', 'transit',
       'access', 'interaction', 'house_rules', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_is_superhost',
       'host_thumbnail_url', 'host_picture_url', 'host_neighborhood',
       'host_verifications', 'host_has_profile_pic', 'host_identity_verified',
       'street', 'neighborhood', 'city', 'suburb', 'state', 'zipcode',
       'smart_location', 'country_code', 'country', 'latitude', 'longitude',
       'is_location_exact', 'property_type', 'room_type', 'accommodates',
       'bathrooms', 'bedrooms', 'beds', 'bed_type', 'amenities', 'price',
       'weekly_price', 'monthly_price', 'security_deposit', 'cleaning_fee',
       'guests_included', 'extra_people', 'minimum_nights', 'maximum_nights',
       'calendar_updated', 

In [ ]:
len(airbnb_df)

22895

## Ejercicio 1 SQL: 

1. Crear una base de datos en SQLite utilizando la libreria [SQLalchemy](https://stackoverflow.com/questions/2268050/execute-sql-from-file-in-sqlalchemy).
https://docs.sqlalchemy.org/en/14/core/engines.html#sqlite

2. Ingestar los datos provistos en 'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/melb_data.csv' en una tabla y el dataset generado en clase con datos de airbnb y sus precios por codigo postal en otra.

3. Implementar consultas en SQL que respondan con la siguiente información:

    - cantidad de registros totales por ciudad.
    - cantidad de registros totales por barrio y ciudad.

4. Combinar los datasets de ambas tablas ingestadas utilizando el comando JOIN de SQL  para obtener un resultado similar a lo realizado con Pandas en clase.  



In [ ]:
#Creamos una base de datos en SQLite
#Agregamos primero la base de datos de Melbourne
engine = create_engine('sqlite:///melb_data.sqlite3', echo=True)
melb_df.to_sql('melb_data', con=engine, if_exists="replace")

2022-07-11 21:00:06,424 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("melb_data")
2022-07-11 21:00:06,428 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-11 21:00:06,436 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("melb_data")
2022-07-11 21:00:06,441 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-11 21:00:06,455 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-11 21:00:06,462 INFO sqlalchemy.engine.Engine 
CREATE TABLE melb_data (
	"index" BIGINT, 
	"Suburb" TEXT, 
	"Address" TEXT, 
	"Rooms" BIGINT, 
	"Type" TEXT, 
	"Price" FLOAT, 
	"Method" TEXT, 
	"SellerG" TEXT, 
	"Date" TEXT, 
	"Distance" FLOAT, 
	"Postcode" FLOAT, 
	"Bedroom2" FLOAT, 
	"Bathroom" FLOAT, 
	"Car" FLOAT, 
	"Landsize" FLOAT, 
	"BuildingArea" FLOAT, 
	"YearBuilt" FLOAT, 
	"CouncilArea" TEXT, 
	"Lattitude" FLOAT, 
	"Longtitude" FLOAT, 
	"Regionname" TEXT, 
	"Propertycount" FLOAT
)


2022-07-11 21:00:06,467 INFO sqlalchemy.engine.Engine [no key 0.00569s] ()
2022-07-11 21:00:06,484 INFO sqlal

In [ ]:
#Ahora le agregamos la base de datos de airbnb
airbnb_df.to_sql('airbnb', con=engine, if_exists="replace")

2022-07-11 21:00:08,012 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("airbnb")
2022-07-11 21:00:08,014 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-11 21:00:08,021 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("airbnb")
2022-07-11 21:00:08,026 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-11 21:00:08,051 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-11 21:00:08,071 INFO sqlalchemy.engine.Engine 
CREATE TABLE airbnb (
	"index" BIGINT, 
	id BIGINT, 
	listing_url TEXT, 
	scrape_id FLOAT, 
	last_scraped TEXT, 
	name TEXT, 
	summary TEXT, 
	space TEXT, 
	description TEXT, 
	neighborhood_overview TEXT, 
	notes TEXT, 
	transit TEXT, 
	access TEXT, 
	interaction TEXT, 
	house_rules TEXT, 
	picture_url TEXT, 
	host_id BIGINT, 
	host_url TEXT, 
	host_name TEXT, 
	host_since TEXT, 
	host_location TEXT, 
	host_about TEXT, 
	host_response_time TEXT, 
	host_response_rate TEXT, 
	host_is_superhost TEXT, 
	host_thumbnail_url TEXT, 
	host_picture_url TEXT, 
	host_neig

Implementamos consultas en SQL que respondan a la siguiente información: **cantidad de registros totales por ciudad.**

Identificamos las ciudades registradas en el dataset de airbnb.

In [ ]:
airbnb_df["city"].unique()

array(['Manningham', 'Moreland', 'Port Phillip', 'Darebin', 'Casey',
       'Melbourne', 'Monash', 'Frankston', 'Yarra', 'Bayside',
       'Boroondara', 'Kingston', 'Stonnington', 'Glen Eira', 'Maroondah',
       'Maribyrnong', 'Yarra Ranges', 'Banyule', 'Whitehorse', 'Melton',
       'Brimbank', 'Hobsons Bay', 'Wyndham', 'Moonee Valley',
       'Whittlesea', 'Greater Dandenong', 'Nillumbik', 'Knox', 'Cardinia',
       'Hume'], dtype=object)

Aplicamos la herramienta SQL para observar la cantidad de registros totales por ciudad en el dataset de airbnb

In [ ]:
# cantidad de registros por ciudad de la base de datos de airbnb
query1 = """SELECT city, COUNT(city) FROM  airbnb 
GROUP BY city
ORDER BY COUNT(city) DESC """


In [ ]:
with engine.connect() as con:
      rs = con.execute(query1)
      print(query1)
      for row in rs:
          print(row)

2022-07-11 21:00:19,980 INFO sqlalchemy.engine.Engine SELECT city, COUNT(city) FROM  airbnb 
GROUP BY city
ORDER BY COUNT(city) DESC 
2022-07-11 21:00:19,985 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT city, COUNT(city) FROM  airbnb 
GROUP BY city
ORDER BY COUNT(city) DESC 
('Melbourne', 7368)
('Port Phillip', 2808)
('Yarra', 2049)
('Stonnington', 1621)
('Moreland', 967)
('Yarra Ranges', 771)
('Darebin', 698)
('Boroondara', 664)
('Glen Eira', 631)
('Whitehorse', 614)
('Monash', 571)
('Maribyrnong', 436)
('Wyndham', 426)
('Bayside', 375)
('Moonee Valley', 344)
('Manningham', 313)
('Kingston', 309)
('Hobsons Bay', 239)
('Banyule', 203)
('Frankston', 177)
('Knox', 175)
('Hume', 170)
('Casey', 153)
('Greater Dandenong', 147)
('Whittlesea', 137)
('Cardinia', 123)
('Maroondah', 115)
('Brimbank', 108)
('Melton', 95)
('Nillumbik', 88)


Podemos observar una mayor cantidad de registros de la ciudad de Melbourne (7368), con respecto a las demas ciudades.

A continuación, observamos el registro de ciudades correspondiente al dataset melb_df. 

In [ ]:
melb_df["CouncilArea"].unique()

array(['Yarra', 'Moonee Valley', 'Port Phillip', 'Darebin', 'Hobsons Bay',
       'Stonnington', 'Boroondara', 'Monash', 'Glen Eira', 'Whitehorse',
       'Maribyrnong', 'Bayside', 'Moreland', 'Manningham', 'Banyule',
       'Melbourne', 'Kingston', 'Brimbank', 'Hume', nan, 'Knox',
       'Maroondah', 'Casey', 'Melton', 'Greater Dandenong', 'Nillumbik',
       'Whittlesea', 'Frankston', 'Macedon Ranges', 'Yarra Ranges',
       'Wyndham', 'Cardinia', 'Unavailable', 'Moorabool'], dtype=object)

Aplicando nuevamente la herramienta SQL, en este caso para el dataset de melb_data.

In [ ]:
# cantidad de registros por ciudad de la base de datos de melb_data
query2 = """SELECT CouncilArea, COUNT(CouncilArea) FROM  melb_data 
GROUP BY CouncilArea
ORDER BY COUNT(CouncilArea) DESC """

In [ ]:
with engine.connect() as con:
      rs = con.execute(query2)
      print(query2)
      for row in rs:
          print(row)

2022-07-11 21:00:20,108 INFO sqlalchemy.engine.Engine SELECT CouncilArea, COUNT(CouncilArea) FROM  melb_data 
GROUP BY CouncilArea
ORDER BY COUNT(CouncilArea) DESC 
2022-07-11 21:00:20,112 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT CouncilArea, COUNT(CouncilArea) FROM  melb_data 
GROUP BY CouncilArea
ORDER BY COUNT(CouncilArea) DESC 
('Moreland', 1163)
('Boroondara', 1160)
('Moonee Valley', 997)
('Darebin', 934)
('Glen Eira', 848)
('Stonnington', 719)
('Maribyrnong', 692)
('Yarra', 647)
('Port Phillip', 628)
('Banyule', 594)
('Bayside', 489)
('Melbourne', 470)
('Hobsons Bay', 434)
('Brimbank', 424)
('Monash', 333)
('Manningham', 311)
('Whitehorse', 304)
('Kingston', 207)
('Whittlesea', 167)
('Hume', 164)
('Wyndham', 86)
('Knox', 80)
('Maroondah', 80)
('Melton', 66)
('Frankston', 53)
('Greater Dandenong', 52)
('Casey', 38)
('Nillumbik', 36)
('Yarra Ranges', 18)
('Cardinia', 8)
('Macedon Ranges', 7)
('Moorabool', 1)
('Unavailable', 1)
(None, 0)


Comparando los resultados obtenidos para ambos conjunto de datos, podemos decir que:


*   Existe una clara diferencia de registros de la ciudad de Melbourne en ambos datasets, observando que la cantidad de registros de esta ciudad en el conjunto de datos de airbnb es de 7368, mientras que en el dataset de melb_df, dicha cantidad es mucho menor (470 registros). Lo mismo ocurre en menor medida para las ciudadades de: Port Phillip, Yarra y Stonnington. 
*   Ademas, se puede observar que hay una mayor cantidad de registros totales en el conjunto de datos de airbnb.



Implementamos consultas en SQL que respondan a la siguiente información:
**cantidad de registros totales por barrio y ciudad.**



In [ ]:
airbnb_df["neighborhood"].unique()

array(['Balwyn North', 'Brunswick', 'St Kilda', 'Thornbury',
       'St Kilda East', nan, 'Richmond', 'Central Business District',
       'Brighton', 'Kew', 'Fitzroy', 'Clifton Hill', 'Northcote',
       'Port Melbourne', 'South Melbourne', 'Windsor', 'Parkville',
       'South Yarra', 'Yarraville/Seddon', 'Elwood', 'Middle Park',
       'Hawthorn', 'Carlton', 'Caulfield', 'Coburg', 'North Melbourne',
       'Prahran', 'East Melbourne', 'Albert Park', 'Collingwood',
       'Abbotsford', 'Spotswood/South Kingsville', 'Armadale',
       'Southbank', 'Balwyn', 'Balaclava/Ripponlea', 'Toorak',
       'Malvern East', 'Flemington', 'Prahran/Windsor', 'Footscray',
       'Preston', 'Carnegie', 'Ashburton', 'Fitzroy North', 'Alphington',
       'Docklands', 'Glen Iris', 'Ormond/McKinnon', 'South Wharf',
       'Bentleigh', 'Newport', 'Carlton North', 'West Melbourne',
       'Kensington', 'Williamstown', 'Camberwell', 'Burnley', 'Malvern',
       'Canterbury'], dtype=object)

In [ ]:
#Cantidad de registros totales por ciudad y barrio de la base de datos airbnb
query3 = """SELECT city, neighborhood, COUNT(neighborhood) 
FROM  airbnb
GROUP BY city, neighborhood"""

In [ ]:
with engine.connect() as con:
      rs = con.execute(query3)
      print(query3)
      for row in rs:
          print(row)

2022-07-11 21:00:20,210 INFO sqlalchemy.engine.Engine SELECT city, neighborhood, COUNT(neighborhood) 
FROM  airbnb
GROUP BY city, neighborhood
2022-07-11 21:00:20,218 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT city, neighborhood, COUNT(neighborhood) 
FROM  airbnb
GROUP BY city, neighborhood
('Banyule', None, 0)
('Banyule', 'Alphington', 16)
('Banyule', 'Preston', 8)
('Banyule', 'Thornbury', 5)
('Bayside', None, 0)
('Bayside', 'Bentleigh', 1)
('Bayside', 'Brighton', 203)
('Bayside', 'Elwood', 1)
('Boroondara', None, 0)
('Boroondara', 'Ashburton', 14)
('Boroondara', 'Balwyn', 70)
('Boroondara', 'Balwyn North', 40)
('Boroondara', 'Camberwell', 82)
('Boroondara', 'Canterbury', 24)
('Boroondara', 'Glen Iris', 32)
('Boroondara', 'Hawthorn', 261)
('Boroondara', 'Kew', 109)
('Boroondara', 'Malvern', 2)
('Brimbank', None, 0)
('Cardinia', None, 0)
('Casey', None, 0)
('Darebin', None, 0)
('Darebin', 'Alphington', 22)
('Darebin', 'Coburg', 4)
('Darebin', 'Northcote', 303)
('Darebin', 'Prest

In [ ]:
melb_df["Suburb"].unique()

array(['Abbotsford', 'Airport West', 'Albert Park', 'Alphington',
       'Altona', 'Altona North', 'Armadale', 'Ascot Vale', 'Ashburton',
       'Ashwood', 'Avondale Heights', 'Balaclava', 'Balwyn',
       'Balwyn North', 'Bentleigh', 'Bentleigh East', 'Box Hill',
       'Braybrook', 'Brighton', 'Brighton East', 'Brunswick',
       'Brunswick West', 'Bulleen', 'Burwood', 'Camberwell', 'Canterbury',
       'Carlton North', 'Carnegie', 'Caulfield', 'Caulfield North',
       'Caulfield South', 'Chadstone', 'Clifton Hill', 'Coburg',
       'Coburg North', 'Collingwood', 'Doncaster', 'Eaglemont',
       'Elsternwick', 'Elwood', 'Essendon', 'Essendon North', 'Fairfield',
       'Fitzroy', 'Fitzroy North', 'Flemington', 'Footscray', 'Glen Iris',
       'Glenroy', 'Gowanbrae', 'Hadfield', 'Hampton', 'Hampton East',
       'Hawthorn', 'Heidelberg Heights', 'Heidelberg West', 'Hughesdale',
       'Ivanhoe', 'Kealba', 'Keilor East', 'Kensington', 'Kew',
       'Kew East', 'Kooyong', 'Maidstone', 

In [ ]:
#Cantidad de registros totales por ciudad y barrio de la base de datos melb_data
query4 = """SELECT CouncilArea, Suburb, COUNT(Suburb) 
FROM  melb_data
GROUP BY CouncilArea, Suburb"""

In [ ]:
with engine.connect() as con:
      rs = con.execute(query4)
      print(query4)
      for row in rs:
          print(row)

2022-07-11 21:00:20,443 INFO sqlalchemy.engine.Engine SELECT CouncilArea, Suburb, COUNT(Suburb) 
FROM  melb_data
GROUP BY CouncilArea, Suburb
2022-07-11 21:00:20,446 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT CouncilArea, Suburb, COUNT(Suburb) 
FROM  melb_data
GROUP BY CouncilArea, Suburb
(None, 'Abbotsford', 1)
(None, 'Aberfeldie', 4)
(None, 'Airport West', 6)
(None, 'Albanvale', 2)
(None, 'Albert Park', 5)
(None, 'Albion', 3)
(None, 'Alphington', 2)
(None, 'Altona', 5)
(None, 'Altona Meadows', 2)
(None, 'Altona North', 7)
(None, 'Ardeer', 1)
(None, 'Armadale', 9)
(None, 'Ascot Vale', 5)
(None, 'Ashburton', 7)
(None, 'Ashwood', 1)
(None, 'Aspendale', 3)
(None, 'Aspendale Gardens', 2)
(None, 'Attwood', 1)
(None, 'Avondale Heights', 1)
(None, 'Balaclava', 2)
(None, 'Balwyn', 6)
(None, 'Balwyn North', 20)
(None, 'Bayswater', 2)
(None, 'Bayswater North', 3)
(None, 'Beaumaris', 7)
(None, 'Bellfield', 1)
(None, 'Bentleigh', 8)
(None, 'Bentleigh East', 15)
(None, 'Berwick', 2)
(None, 

**Combinamos los datasets de ambas tablas ingestadas utilizando el comando JOIN de SQL para obtener un resultado similar a lo realizado con Pandas en clase.**

In [ ]:
with engine.connect() as con:
      rs = con.execute("SELECT monthly_price FROM airbnb WHERE monthly_price IS NOT null")
      print()
      for row in rs:
          print(row)

2022-07-11 21:00:20,710 INFO sqlalchemy.engine.Engine SELECT monthly_price FROM airbnb WHERE monthly_price IS NOT null
2022-07-11 21:00:20,712 INFO sqlalchemy.engine.Engine [raw sql] ()

(803.0,)
(4452.0,)
(920.0,)
(1900.0,)
(1400.0,)
(1540.0,)
(5572.0,)
(5100.0,)
(1150.0,)
(1000.0,)
(1100.0,)
(4172.0,)
(3000.0,)
(4000.0,)
(1260.0,)
(1500.0,)
(800.0,)
(8648.0,)
(3612.0,)
(1900.0,)
(2800.0,)
(657.0,)
(4900.0,)
(750.0,)
(2400.0,)
(3892.0,)
(980.0,)
(4300.0,)
(3400.0,)
(3400.0,)
(1300.0,)
(1200.0,)
(2250.0,)
(3862.0,)
(2320.0,)
(2500.0,)
(2000.0,)
(2500.0,)
(3300.0,)
(3400.0,)
(3892.0,)
(3600.0,)
(3200.0,)
(15000.0,)
(1250.0,)
(3436.0,)
(2700.0,)
(1080.0,)
(3850.0,)
(4200.0,)
(5572.0,)
(3632.0,)
(2500.0,)
(4000.0,)
(1800.0,)
(1900.0,)
(6245.0,)
(3185.0,)
(3750.0,)
(1775.0,)
(9000.0,)
(1450.0,)
(4410.0,)
(1960.0,)
(2800.0,)
(2900.0,)
(4500.0,)
(2436.0,)
(4410.0,)
(4867.0,)
(2252.0,)
(3000.0,)
(975.0,)
(3763.0,)
(5931.0,)
(800.0,)
(3800.0,)
(5300.0,)
(900.0,)
(3900.0,)
(4715.0,)
(4258.0,)
(

In [ ]:
query5="""SELECT CAST(m.postcode AS int), CAST(a.zipcode AS int), a.price, a.weekly_price, a.monthly_price, m.price
FROM melb_data m
INNER JOIN airbnb a
ON m.postcode=a.zipcode
LIMIT 1000
"""

In [ ]:
with engine.connect() as con:
      rs = con.execute(query5)
      print(query5)
      for row in rs:
          print(row)

2022-07-11 21:00:21,840 INFO sqlalchemy.engine.Engine SELECT CAST(m.postcode AS int), CAST(a.zipcode AS int), a.price, a.weekly_price, a.monthly_price, m.price
FROM melb_data m
INNER JOIN airbnb a
ON m.postcode=a.zipcode
LIMIT 1000

2022-07-11 21:00:21,841 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT CAST(m.postcode AS int), CAST(a.zipcode AS int), a.price, a.weekly_price, a.monthly_price, m.price
FROM melb_data m
INNER JOIN airbnb a
ON m.postcode=a.zipcode
LIMIT 1000

(3105, 3105, 60, None, None, 599900.0)
(3105, 3105, 60, None, None, 760000.0)
(3105, 3105, 60, None, None, 787000.0)
(3105, 3105, 60, None, None, 794000.0)
(3105, 3105, 60, None, None, 812500.0)
(3105, 3105, 60, None, None, 833000.0)
(3105, 3105, 60, None, None, 851000.0)
(3105, 3105, 60, None, None, 860000.0)
(3105, 3105, 60, None, None, 880000.0)
(3105, 3105, 60, None, None, 890000.0)
(3105, 3105, 60, None, None, 903000.0)
(3105, 3105, 60, None, None, 905000.0)
(3105, 3105, 60, None, None, 910000.0)
(3105, 3105, 6

## Ejercicio 2: 

1. Seleccionar un subconjunto de columnas que les parezcan relevantes al problema de predicción del valor de la propiedad. Justificar las columnas seleccionadas y las que no lo fueron.
 - Eliminar los valores extremos que no sean relevantes para la predicción de valores de las propiedades.

2. Agregar información adicional respectiva al entorno de una propiedad a partir del [conjunto de datos de AirBnB](https://www.kaggle.com/tylerx/melbourne-airbnb-open-data?select=cleansed_listings_dec18.csv) utilizado en el práctico. 
  1. Seleccionar qué variables agregar y qué combinaciones aplicar a cada una. Por ejemplo, pueden utilizar solo la columna `price`, o aplicar múltiples transformaciones como la mediana o el mínimo.
  1. Utilizar la variable zipcode para unir los conjuntos de datos. Sólo incluir los zipcodes que tengan una cantidad mínima de registros (a elección) como para que la información agregada sea relevante.
  2. Investigar al menos otras 2 variables que puedan servir para combinar los datos, y justificar si serían adecuadas o no. Pueden asumir que cuentan con la ayuda de anotadores expertos para encontrar equivalencias entre barrios o direcciones, o que cuentan con algoritmos para encontrar las n ubicaciones más cercanas a una propiedad a partir de sus coordenadas geográficas. **NO** es necesario que realicen la implementación.

Pueden leer otras columnas del conjunto de AirBnB además de las que están en `interesting_cols`, si les parecen relevantes.

Primero obtenemos la descripción de todas las columnas:

Rooms: Number of rooms

Price: Price in dollars

Method: S - property sold; SP - property sold prior; PI - property passed in; PN - sold prior not disclosed; SN - sold not disclosed; NB - no bid; VB - vendor bid; W - withdrawn prior to auction; SA - sold after auction; SS - sold after auction price not disclosed. N/A - price or highest bid not available.

Type: br - bedroom(s); h - house,cottage,villa, semi,terrace; u - unit, duplex; t - townhouse; dev site - development site; o res - other residential.

SellerG: Real Estate Agent

Date: Date sold

Distance: Distance from CBD

Regionname: General Region (West, North West, North, North east …etc)

Propertycount: Number of properties that exist in the suburb.

Bedroom2 : Scraped # of Bedrooms (from different source)

Bathroom: Number of Bathrooms

Car: Number of carspots

Landsize: Land Size

BuildingArea: Building Size

CouncilArea: Governing council for the area

Las columnas que consideramos importantes para una futura predicción del precio de una propiedad son:
- Rooms, Bedroom2 y bathrooms: Se escogieron estas tres variables ya que se considera que cuanto mayor es la cantidad de habitaciones/baños que tenga la casa, mayor será el precio de la misma. 
- Distance: 
- Landsize: se considera importante esta variable para poder analizar la incidencia del tamaño del terreno el el precio de la casa. 
- BuildingArea: La superficie de edificación sin duda sera una variable que incida en el precio final de una casa. 
- Car: 
- YearBuilt: se elige la variable Yearbuilt ya que se presupone una relación entre la edad de la casa y su precio. 


Y el Precio para evaluar futuras predicciones.

Para filtrar los valores extremos **Numericos** utilizamos el método de los 3 sigmas: filtramos los valores por encima o por debajo de 3 desvios estandar de la media.

Para filtrar los valores extremos **Categóricos**:
Para cada variable, ponderamos cada categoria calculando las frecuencias relativas. Luego, eliminamos los registros de aquellas categorias que tenian menos frecuencia relativa que un *0.5%*.

In [ ]:
melb_df = melbourne_df.copy()

In [ ]:
def filter_outliers_mask(df, cols, n_desvios_ = 3, verbosity = True):

  df_size = df.shape[0]

  idx_to_delete=[]

  for c in cols:
    print(c)
    media = df[c].mean()
    desvio = df[c].std()
    print(desvio/media)
    
    minimo, maximo = (media - n_desvios * desvio, media + n_desvios * desvio)
    
    mask = (df[c] < minimo) | (df[c] > maximo)

    idx_to_delete.append(list(df.loc[mask].index))
    
  idx_to_delete = numpy.array([item for sublist in idx_to_delete for item in sublist])

  if verbosity:
    print("DF Size", df_size)
    print(f"Eliminaremos un {numpy.round(len(idx_to_delete)/df_size * 100,1)}% de los registros")
    print("DF Processed size: ", df_size - len(idx_to_delete))

  
  return idx_to_delete

def filter_df(df, idx_to_delete):
  return df.drop(index = idx_to_delete)



def filter_categorical(df, cols, perc_of_cut = 0.005, verbosity = True):  
  df_size = df.shape[0]
  idx_to_delete=[]  

  for c in categorical_cols:   
    categories_count = df[c].value_counts()/df[c].count()
    categories_to_delete = categories_count[categories_count < perc_of_cut].index
    idx_to_delete_categorical = df[df[c].isin(categories_to_delete)].index
    idx_to_delete.append(idx_to_delete_categorical)
 
 
  idx_to_delete = numpy.array([item for sublist in idx_to_delete for item in sublist])

  if verbosity:
    print("DF Size", df_size)
    print(f"Eliminaremos un {numpy.round(len(idx_to_delete)/df_size * 100,1)}% de los registros")
    print("DF Processed size: ", df_size - len(idx_to_delete))

  
  return idx_to_delete

In [ ]:
categorical_cols = ['Type','CouncilArea']
numerical_cols = ['Rooms',  'Bedroom2', 'Bathroom','Distance',  'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'Price']
n_desvios = 3

# Numerical filtering 3 sigma
print("Numerical\n")
idx_to_delete_numerical = filter_outliers_mask(melb_df, numerical_cols, n_desvios_ = n_desvios, verbosity = True)

# Categorical filtering
print("\n Categorical")
idx_to_delete_categorical = filter_categorical(melb_df, categorical_cols, perc_of_cut = 0.001, verbosity = True)

idx_to_delete = numpy.append(idx_to_delete_numerical, idx_to_delete_categorical)
idx_to_delete = numpy.unique(idx_to_delete)

Numerical

Rooms
0.32530595528008216
Bedroom2
0.33139326001719105
Bathroom
0.45084930122347244
Distance
0.5788966792346524
Car
0.5978809976183062
Landsize
7.146407581386584
BuildingArea
3.5600638558653595
YearBuilt
0.018971884592221013
Price
0.5943294471922271
DF Size 13580
Eliminaremos un 7.2% de los registros
DF Processed size:  12602

 Categorical
DF Size 13580
Eliminaremos un 0.1% de los registros
DF Processed size:  13563


In [ ]:
n_shape_inicial = melb_df.shape[0]
melb_df = filter_df(melb_df, idx_to_delete)

for c in numerical_cols:
  print(c)
  media = melb_df[c].mean()
  desvio = melb_df[c].std()
  print(desvio/media)

print(n_shape_inicial, '\n',melb_df.shape[0])
print(melb_df.shape[0] / n_shape_inicial)
      

Rooms
0.31322108880705063
Bedroom2
0.31325213104609373
Bathroom
0.4125850351983081
Distance
0.501147316698878
Car
0.5482800388684067
Landsize
1.241275257278208
BuildingArea
0.5832717349024185
YearBuilt
0.018381400091815128
Price
0.5042109126071634
13580 
 12784
0.9413843888070692


**Seleccionamos qué variables agregar y qué combinaciones aplicar a cada una.**

En principio, agregaríamos las columnas: 
1.   price = "precio de alquiler por día" 
2.   weekly_price = "precio de alquiler por semana"
3.   monthly_price = "precio de alquiler por mes"
4.   zipcode = "Código postal"

con las siguientes transformaciones: 

1.   El valor mínimo, máximo y promedio de los precios
2.   El conteo de registros por cada zipcode ("zipcode_count")






In [ ]:
airbnb_df.zipcode.value_counts()[:5]

3000.0    2491
3000       876
3006.0     860
3182.0     579
3182       556
Name: zipcode, dtype: int64

In [ ]:
airbnb_df['zipcode'] = pandas.to_numeric(airbnb_df.zipcode, errors='coerce')

In [ ]:
airbnb_df.zipcode.value_counts().head(5)

3000.0    3367
3006.0    1268
3182.0    1135
3141.0     881
3004.0     728
Name: zipcode, dtype: int64

In [ ]:
relevant_cols = ['zipcode', 'price', 'weekly_price', 'monthly_price']
## Option 1
# Pass a dictionary where the keys are the original columns to aggregate and
# the values are the operations (or list of operations).
airbnb_price_by_zipcode = airbnb_df[relevant_cols].groupby('zipcode')\
  .agg({'price': ['min', 'mean', "max"], 'weekly_price': ['min', 'mean', "max"],
        'monthly_price': ['min', 'mean', "max"], "zipcode": "count"})\
  .reset_index()
# Flatten the two level columns
airbnb_price_by_zipcode.columns = [
  ' '.join(col).strip()
  for col in airbnb_price_by_zipcode.columns.values]
# Rename columns
airbnb_price_by_zipcode = airbnb_price_by_zipcode.rename(
    columns={'price min': "airbnb_price_min",
             'price mean': 'airbnb_price_mean',
             'price max':'airbnb_price_max',
             'weekly_price min': 'airbnb_weekly_price_min',
             'weekly_price mean': 'airbnb_weekly_price_mean',
             'weekly_price max': 'airbnb_weekly_price_max',
             'monthly_price min': 'airbnb_monthly_price_min',
             'monthly_price mean': 'airbnb_monthly_price_mean',
             'monthly_price max': 'airbnb_monthly_price_max',
             'zipcode count': 'zipcode_count',
                }
)
airbnb_price_by_zipcode



zipcode  airbnb_price_min  airbnb_price_mean  airbnb_price_max  \
0     2010.0                40          40.000000                40   
1     2134.0                50          50.000000                50   
2     2582.0               104         104.000000               104   
3     3000.0                 0         150.504307              9000   
4     3001.0               120         132.500000               145   
..       ...               ...                ...               ...   
242   7367.0               230         230.000000               230   
243   8938.0               296         296.000000               296   
244   9004.0               226         226.000000               226   
245   9540.0               245         245.000000               245   
246  30122.0                75          75.000000                75   

     airbnb_weekly_price_min  airbnb_weekly_price_mean  \
0                        NaN                       NaN   
1                        NaN                       NaN   
2                        NaN                       NaN   
3                      150.0                918.738956   
4                        NaN                       NaN   
..                       ...                       ...   
242                      NaN                       NaN   
243                      NaN                       NaN   
244                      NaN                       NaN   
245                      NaN                       NaN   
246                    525.0                525.000000   

     airbnb_weekly_price_max  airbnb_monthly_price_min  \
0                        NaN                       NaN   
1                        NaN                       NaN   
2                        NaN                       NaN   
3                     3325.0                     550.0   
4                        NaN                       NaN   
..                       ...                       ...   
242                      NaN                       NaN   
243                      NaN                       NaN   
244                      NaN                       NaN   
245                      NaN                       NaN   
246                    525.0                    2281.0   

     airbnb_monthly_price_mean  airbnb_monthly_price_max  zipcode_count  
0                          NaN                       NaN              1  
1                          NaN                       NaN              1  
2                          NaN                       NaN              1  
3                  3407.204651                    8060.0           3367  
4                          NaN                       NaN              2  
..                         ...                       ...            ...  
242                        NaN                       NaN              1  
243                        NaN                       NaN              1  
244                        NaN                       NaN              1  
245                        NaN                       NaN              1  
246                2281.000000                    2281.0              1  

[247 rows x 11 columns]

**Utilizamos la variable zipcode para unir los conjuntos de datos.** Sólo incluimos los zipcodes que tengan una cantidad mínima de registros como para que la información agregada sea relevante.

Seleccionamos como cantidad mínima 25 registros (valor arbitrario cercano a la mediana).

In [ ]:
airbnb_price_by_zipcode["zipcode_count"].describe()

count     247.000000
mean       92.101215
std       261.914701
min         1.000000
25%         8.500000
50%        27.000000
75%        73.000000
max      3367.000000
Name: zipcode_count, dtype: float64

In [ ]:
n_registros = 25

airbnb_price_by_zipcode_min= airbnb_price_by_zipcode[airbnb_price_by_zipcode["zipcode_count"] >= n_registros]
airbnb_price_by_zipcode_min

zipcode  airbnb_price_min  airbnb_price_mean  airbnb_price_max  \
3     3000.0                 0         150.504307              9000   
5     3002.0                19         200.979695              1550   
6     3003.0                19         130.205993               550   
7     3004.0                12         158.271978              1771   
8     3006.0                17         188.851735              2998   
..       ...               ...                ...               ...   
206   3788.0                86         249.370787               610   
214   3799.0                35         196.932584               750   
218   3805.0                30         225.235294              1101   
219   3806.0                37         488.272727             12501   
233   3977.0                18          92.955556               341   

     airbnb_weekly_price_min  airbnb_weekly_price_mean  \
3                      150.0                918.738956   
5                      140.0                956.379310   
6                      362.0                760.200000   
7                      298.0               1061.171429   
8                      150.0               1239.827273   
..                       ...                       ...   
206                    600.0               1547.500000   
214                    173.0                992.666667   
218                    380.0                622.000000   
219                    540.0                540.000000   
233                    198.0                422.000000   

     airbnb_weekly_price_max  airbnb_monthly_price_min  \
3                     3325.0                     550.0   
5                     2030.0                     700.0   
6                     1388.0                    1350.0   
7                     3200.0                    1000.0   
8                     3500.0                    1500.0   
..                       ...                       ...   
206                   2445.0                    2000.0   
214                   1740.0                     609.0   
218                   1106.0                    1400.0   
219                    540.0                       NaN   
233                   1000.0                     950.0   

     airbnb_monthly_price_mean  airbnb_monthly_price_max  zipcode_count  
3                  3407.204651                    8060.0           3367  
5                  3712.714286                    8000.0            197  
6                  2629.500000                    4689.0            267  
7                  3759.925926                    8998.0            728  
8                  4358.556701                   15000.0           1268  
..                         ...                       ...            ...  
206                5441.500000                    9000.0             89  
214                2194.333333                    4000.0             89  
218                2178.666667                    3736.0             34  
219                        NaN                       NaN             33  
233                1025.000000                    1100.0             45  

[128 rows x 11 columns]

In [ ]:
merged_sales_df = melb_df.merge(
    airbnb_price_by_zipcode_min, how='left',
    left_on='Postcode', right_on='zipcode'
)
print("% de registros no intersectados ", numpy.round(merged_sales_df.zipcode.isnull().sum() / merged_sales_df.shape[0] * 100, 1), "%")
merged_sales_df.sample(10)

% de registros no intersectados  13.1 %


Suburb                   Address  Rooms Type      Price Method  \
1231        Brunswick             215/1 Dods St      3    u  1044000.0      S   
2903          Hampton           118 Ludstone St      2    h  1615000.0     SA   
5255          Rosanna           6/3 Prospect Rd      2    u   452000.0      S   
5894     Surrey Hills         706 Canterbury Rd      2    h   680000.0     PI   
11675           Lalor           123 Gillwell Rd      4    h   715000.0      S   
4544   Port Melbourne             516/99 Dow St      2    u   820000.0     VB   
4610          Prahran                16 Park Rd      3    t  1245000.0     PI   
1537       Canterbury              33 Myrtle Rd      3    h  1587500.0      S   
10763     Craigieburn  28/60 Cradle Mountain Dr      2    t   310000.0     PI   
7202        Bellfield              125 Oriel Rd      3    h   894000.0      S   

        SellerG        Date  Distance  Postcode  ...  airbnb_price_min  \
1231     Jellis  16/07/2016       5.2    3056.0  ...              25.0   
2903   Chisholm  12/11/2016      13.7    3188.0  ...              35.0   
5255     Nelson  28/08/2016      11.4    3084.0  ...              24.0   
5894       Noel  10/12/2016      11.2    3127.0  ...               0.0   
11675     Barry   3/09/2017      16.3    3075.0  ...               NaN   
4544     Cayzer  14/05/2016       3.8    3207.0  ...               0.0   
4610   Marshall   6/08/2016       4.5    3181.0  ...              22.0   
1537   Marshall   3/12/2016       9.0    3126.0  ...               NaN   
10763       Ray  15/07/2017      20.6    3064.0  ...              12.0   
7202        Ray   8/04/2017       9.5    3081.0  ...              25.0   

       airbnb_price_mean  airbnb_price_max  airbnb_weekly_price_min  \
1231           94.672646             599.0                    170.0   
2903          245.112676             898.0                    250.0   
5255          127.208333             732.0                    250.0   
5894           86.411765             260.0                    390.0   
11675                NaN               NaN                      NaN   
4544          200.210117            1150.0                    360.0   
4610          148.169565            1402.0                    250.0   
1537                 NaN               NaN                      NaN   
10763          71.404762             251.0                    180.0   
7202           83.657895             453.0                    200.0   

       airbnb_weekly_price_mean  airbnb_weekly_price_max  \
1231                 618.758065                   2700.0   
2903                1273.600000                   5768.0   
5255                1128.750000                   3100.0   
5894                 444.500000                    499.0   
11675                       NaN                      NaN   
4544                1480.790698                   7000.0   
4610                 868.148148                   2730.0   
1537                        NaN                      NaN   
10763                185.000000                    190.0   
7202                 291.428571                    400.0   

      airbnb_monthly_price_min  airbnb_monthly_price_mean  \
1231                     600.0                2308.978723   
2903                     900.0                4694.555556   
5255                     900.0                2298.000000   
5894                    1699.0                1699.000000   
11675                      NaN                        NaN   
4544                    1850.0                4743.964286   
4610                     860.0                3304.382353   
1537                       NaN                        NaN   
10763                      NaN                        NaN   
7202                     600.0                 937.500000   

       airbnb_monthly_price_max zipcode_count  
1231                     4590.0         446.0  
2903                    18663.0          71.0  
5255                     4500.0          72.0  
5

A continuación, analizamos la cantidad de datos faltantes en el dataset merged_sales_df.

In [ ]:
merged_sales_df.columns
print(merged_sales_df.isnull().sum()[merged_sales_df.isnull().sum() > 0])
print(round((merged_sales_df.isnull().sum())/merged_sales_df.shape[0],2)[merged_sales_df.isnull().sum() > 0])

Car                            61
BuildingArea                 6122
YearBuilt                    5081
CouncilArea                  1227
zipcode                      1669
airbnb_price_min             1669
airbnb_price_mean            1669
airbnb_price_max             1669
airbnb_weekly_price_min      1731
airbnb_weekly_price_mean     1731
airbnb_weekly_price_max      1731
airbnb_monthly_price_min     1938
airbnb_monthly_price_mean    1938
airbnb_monthly_price_max     1938
zipcode_count                1669
dtype: int64
Car                          0.00
BuildingArea                 0.48
YearBuilt                    0.40
CouncilArea                  0.10
zipcode                      0.13
airbnb_price_min             0.13
airbnb_price_mean            0.13
airbnb_price_max             0.13
airbnb_weekly_price_min      0.14
airbnb_weekly_price_mean     0.14
airbnb_weekly_price_max      0.14
airbnb_monthly_price_min     0.15
airbnb_monthly_price_mean    0.15
airbnb_monthly_price_max     0.15
z

Observando que los atributos "airbnb_weekly_price" e "airbnb_monthly_price" contienen un 12% y un 14% de datos faltantes respectivamente, se considera necesario eliminar  estos datos del dataframe merged_sales_df para evitar que estos generen complicaciones innecesearias en la etapa de imputación, ya que  se estaría aportando información irrelevante y artificial, lo cual no tiene sentido teniendo en cuenta que estamos realizando una limpieza y curación de los datos. **Es por esto que únicamente nos quedaremos con la columna "airbnb_price_mean", la cual contiene menor cantidad de datos nulos.**

In [ ]:
#Drop columns with missing data
merged_sales_drop_df=merged_sales_df.drop(['airbnb_weekly_price_min', 'airbnb_weekly_price_mean',
                              'airbnb_weekly_price_max','airbnb_monthly_price_min','airbnb_monthly_price_mean',
                              'airbnb_monthly_price_max','airbnb_price_min','airbnb_price_max'], axis=1)

**Investigamos otras variables que podrían servir para combinar los datos.**

Una variable podría ser el barrio, para lo cual determinaríamos (utilizando un algoritmo adecuado) qué latitud y qué longitud corresponde para cada barrio usando el dataset de airbnb. Luego, utilizaríamos esa información para discriminar por barrio el dataset de Melbourne.

Y otra variable más podría ser la distancia de la vivienda a la costa, la cual la determinaríamos también utilizando la latitud y longitud. Esta variable la dividiríamos en intervalos, y pensamos que esta variable podría afectar el precio de la vivienda.

## Ejercicio 3:

Crear y guardar un nuevo conjunto de datos con todas las transformaciones realizadas anteriormente.

In [ ]:
# Read Melb dataframe
melb_df = pandas.read_csv(
    'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/melb_data.csv')
# Read airbnb dataframe
airbnb_df = pandas.read_csv(
    'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/cleansed_listings_dec18.csv')

# Drop rows with outliers using the 3 sigma rule and frequency in categorical variables
categorical_cols = ['Type','CouncilArea']
numerical_cols = ['Rooms',  'Bedroom2', 'Bathroom','Distance',  'Car',
       'Landsize', 'BuildingArea', 'YearBuilt','Price']

cols = numerical_cols + categorical_cols + ["Postcode"]

n_desvios = 3


# Numerical filtering 3 sigma
print("Numerical\n")
idx_to_delete_numerical = filter_outliers_mask(melb_df, numerical_cols, n_desvios_ = n_desvios, verbosity = True)

# Categorical filtering
print("\n Categorical")
idx_to_delete_categorical = filter_categorical(melb_df, categorical_cols, perc_of_cut = 0.001, verbosity = True)

idx_to_delete = numpy.append(idx_to_delete_numerical, idx_to_delete_categorical)
idx_to_delete = numpy.unique(idx_to_delete)

# Melbourne DataFrame
melb_df = filter_df(melb_df, idx_to_delete)
melb_df = melb_df[cols]

airbnb_df['zipcode'] = pandas.to_numeric(airbnb_df.zipcode, errors='coerce')

# Get Zipcode's features from airbnb dataset
relevant_cols = ['zipcode', 'price', 'weekly_price', 'monthly_price']

# Groupby
airbnb_price_by_zipcode = airbnb_df[relevant_cols].groupby('zipcode')\
  .agg({'price': ['min', 'mean', "max"], 'weekly_price': ['min', 'mean', "max"],
        'monthly_price': ['min', 'mean', "max"], "zipcode": "count"})\
  .reset_index()
# Flatten the two level columns
airbnb_price_by_zipcode.columns = [
  ' '.join(col).strip()
  for col in airbnb_price_by_zipcode.columns.values]
# Rename columns
airbnb_price_by_zipcode = airbnb_price_by_zipcode.rename(
    columns={'price min': "airbnb_price_min",
             'price mean': 'airbnb_price_mean',
             'price max':'airbnb_price_max',
             'weekly_price min': 'airbnb_weekly_price_min',
             'weekly_price mean': 'airbnb_weekly_price_mean',
             'weekly_price max': 'airbnb_weekly_price_max',
             'monthly_price min': 'airbnb_monthly_price_min',
             'monthly_price mean': 'airbnb_monthly_price_mean',
             'monthly_price max': 'airbnb_monthly_price_max',
             'zipcode count': 'zipcode_count'
                }
)

# Filter  
n_registros = 25
airbnb_price_by_zipcode_min= airbnb_price_by_zipcode[airbnb_price_by_zipcode["zipcode_count"] >= n_registros] 

# Change type
airbnb_price_by_zipcode_min.loc[:, "zipcode"] = airbnb_price_by_zipcode_min.zipcode.astype(float)
melb_df.loc[:, "Postcode"] = melb_df.Postcode.astype(float)

# Merge
merged_sales_df = melb_df.merge(
    airbnb_price_by_zipcode_min, how='left',
    left_on='Postcode', right_on='zipcode'
)
#Drop columns with missing data
merged_sales_drop_df=merged_sales_df.drop(['airbnb_weekly_price_min', 'airbnb_weekly_price_mean',
                              'airbnb_weekly_price_max','airbnb_monthly_price_min','airbnb_monthly_price_mean',
                              'airbnb_monthly_price_max','airbnb_price_min','airbnb_price_max'], axis=1)

merged_sales_drop_df.to_csv("/content/melb_data_processed.csv", index=False)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (35,77) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)
/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


Numerical

Rooms
0.32530595528008216
Bedroom2
0.33139326001719105
Bathroom
0.45084930122347244
Distance
0.5788966792346524
Car
0.5978809976183062
Landsize
7.146407581386584
BuildingArea
3.5600638558653595
YearBuilt
0.018971884592221013
Price
0.5943294471922271
DF Size 13580
Eliminaremos un 7.2% de los registros
DF Processed size:  12602

 Categorical
DF Size 13580
Eliminaremos un 0.1% de los registros
DF Processed size:  13563


In [ ]:
merged_sales_drop_df.columns
merged_sales_drop_df.isnull().sum()[merged_sales_drop_df.isnull().sum() > 0]

Car                    61
BuildingArea         6122
YearBuilt            5081
CouncilArea          1227
zipcode              1669
airbnb_price_mean    1669
zipcode_count        1669
dtype: int64

In [ ]:
melb_df.Postcode.dtype, airbnb_price_by_zipcode_min.zipcode.dtype

(dtype('float64'), dtype('float64'))

In [ ]:
merged_sales_drop_df.to_csv('merged_sales_df.csv')

from google.colab import files
files.download('merged_sales_df.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
merged_sales_drop_df

Rooms  Bedroom2  Bathroom  Distance  Car  Landsize  BuildingArea  \
0          2       2.0       1.0       2.5  1.0     202.0           NaN   
1          2       2.0       1.0       2.5  0.0     156.0          79.0   
2          3       3.0       2.0       2.5  0.0     134.0         150.0   
3          3       3.0       2.0       2.5  1.0      94.0           NaN   
4          4       3.0       1.0       2.5  2.0     120.0         142.0   
...      ...       ...       ...       ...  ...       ...           ...   
12779      3       3.0       2.0      16.5  2.0     256.0           NaN   
12780      4       4.0       2.0      16.7  2.0     652.0           NaN   
12781      3       3.0       2.0       6.8  2.0     333.0         133.0   
12782      3       3.0       2.0       6.8  4.0     436.0           NaN   
12783      4       4.0       1.0       6.3  1.0     362.0         112.0   

       YearBuilt      Price Type CouncilArea  Postcode  zipcode  \
0            NaN  1480000.0    h       Yarra    3067.0   3067.0   
1         1900.0  1035000.0    h       Yarra    3067.0   3067.0   
2         1900.0  1465000.0    h       Yarra    3067.0   3067.0   
3            NaN   850000.0    h       Yarra    3067.0   3067.0   
4         2014.0  1600000.0    h       Yarra    3067.0   3067.0   
...          ...        ...  ...         ...       ...      ...   
12779        NaN   582000.0    h         NaN    3049.0      NaN   
12780     1981.0  1245000.0    h         NaN    3150.0   3150.0   
12781     1995.0  1031000.0    h         NaN    3016.0   3016.0   
12782     1997.0  1170000.0    h         NaN    3016.0   3016.0   
12783     1920.0  1285000.0    h         NaN    3013.0   3013.0   

       airbnb_price_mean  zipcode_count  
0             130.624031          258.0  
1             130.624031          258.0  
2             130.624031          258.0  
3             130.624031          258.0  
4             130.624031          258.0  
...                  ...            ...  
12779                NaN            NaN  
12780         124.026455          189.0  
12781         191.094595           74.0  
12782         191.094595           74.0  
12783         135.609756           82.0  

[12784 rows x 15 columns]